networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/DiscoveryNetwork.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.networking.p2p;
+
+import java.util.Optional;
+import java.util.stream.Stream;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes;
+import tech.pegasys.artemis.networking.p2p.discovery.ConnectionManager;
+import tech.pegasys.artemis.networking.p2p.discovery.DiscoveryService;
+import tech.pegasys.artemis.networking.p2p.discovery.discv5.DiscV5Service;
+import tech.pegasys.artemis.networking.p2p.network.DelegatingP2PNetwork;
+import tech.pegasys.artemis.networking.p2p.network.NetworkConfig;
+import tech.pegasys.artemis.networking.p2p.network.P2PNetwork;
+import tech.pegasys.artemis.networking.p2p.peer.NodeId;
+import tech.pegasys.artemis.networking.p2p.peer.Peer;
+import tech.pegasys.artemis.networking.p2p.peer.PeerConnectedSubscriber;
+import tech.pegasys.artemis.util.async.DelayedExecutorAsyncRunner;
+import tech.pegasys.artemis.util.async.SafeFuture;
+
+public class DiscoveryNetwork<P extends Peer> extends DelegatingP2PNetwork<P> {
+  private static final Logger LOG = LogManager.getLogger();
+  private final P2PNetwork<P> p2pNetwork;
+  private final DiscoveryService discoveryService;
+  private final ConnectionManager connectionManager;
+
+  public DiscoveryNetwork(
+      final P2PNetwork<P> p2pNetwork,
+      final DiscoveryService discoveryService,
+      final ConnectionManager connectionManager) {
+    super(p2pNetwork);
+    this.p2pNetwork = p2pNetwork;
+    this.discoveryService = discoveryService;
+    this.connectionManager = connectionManager;
```

Comment:
Yeah I'm not sure what the final design will be.  The trouble with a stand-alone `ConnectionManager` is that you wind up with a cycle of dependencies because the `ConnectionManager` needs the `P2PNetwork` but `Eth2P2PNetwork` wants access to the `ConnectionManager`.

We can push the construction of `DiscoveryNetwork` inside `Eth2NetworkBuilder` but it doesn't really change all that much.  It wouldn't surprise me if we do refactor this area, but I'm really not sure what it winds up looking like just yet.  The whole architecture of how topic discovery works seems to still be up in the air a bit.

Reply:
If `ConnectionManager` holds the base `P2PNetwork` (`LibP2PNetwork`) rather than the higher-level `Eth2Network`, you don't get circular dependencies.

Reply:
ok yeah I see how that works.  So for now I'm giving it `DiscoveryNetwork` but `Eth2Network` is still just treating it as any old `P2PNetwork`.  I suspect once we start addressing topic discovery it will wind up taking either explicitly a `DiscoveryNetwork` or some interface that extends `P2PNetwork` but with extra methods for subscribing to topics etc.